### ISSUE TITLE
[BUG] Migration from AppCenter Push not clear at all, installation fails

### ISSUE THREAD

**Issue by andrekiba** [Azure/azure-notificationhubs-xamarin#21]
**Describe the bug**
Seems impossibile to migrate devices from AppCenter Push to NotificationHub without re-installing the application.

***Exception or Stack Trace***
WindowsAzure.Messaging.NotificationHubs.ClientException: Azure Notification Hub rejected the request.
`com.microsoft.windowsazure.messaging.notificationhubs.ClientException: Azure Notification Hub rejected the request.
com.microsoft.windowsazure.messaging.notificationhubs.NotificationHubInstallationAdapter.convertVolleyException NotificationHubInstallationAdapter.java:225
com.microsoft.windowsazure.messaging.notificationhubs.NotificationHubInstallationAdapter$RetrySession.onErrorResponse NotificationHubInstallationAdapter.java:181
com.android.volley.Request.deliverError Request.java:617
com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run ExecutorDelivery.java:104
android.os.Handler.handleCallback Handler.java:790
android.os.Handler.dispatchMessage Handler.java:99
android.os.Looper.loop Looper.java:164
android.app.ActivityThread.main ActivityThread.java:6518
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run RuntimeInit.java:438
com.android.internal.os.ZygoteInit.main ZygoteInit.java:807`

**To Reproduce**
With AppCenter Push to send a notification to a specific device you can use the AppCenter InstallationId.
The concept is the same as the "special" installation id tag in notification hub.
So in our backend we usually have a db table that stores the InstallationId for each device.
Then when we need to send the notification (from backend) is sufficient to retrieve the InstallationId to target the device and use the AppCenter Push REST API to fire the notification. Easy, and it works very good.
Now we have to migrate to NotificationHub...and is unbelievable, for another time after HockeyApp migration, the lack of a clear path and documentation! Every time "you" break something.
So the idea is, use this library (and please, when we will have a nuget package??), save the NotificationHub InstallationId (different than the AppCenter one) on application startup, send this information to the backend and also store the NotificationHub InstallationId in the devices table.
In this way I can still use AppCenter Push (until it is turned off) to send for "old" clients and as soon as the clients are updated (only after update I have the the new NotificationHub InstallationId) then I can use NotificationHub.
The problem is that the NotificationHub Installation always fails in case of in-app update (OnInstallationSaveFailed).
Instead it always works re-installing the application as new (not a possible scenario at all!!).
The code is obviously the same.
Why?

Another thing: why in case of successfully installation saved I can see that the installation has an expiration date when instead the notification hub documentation clearly says that an installation by default never expires??

Please, I can imagine that "you" are not the same team as the AppCenter one...but you can speak to each other??!!
We need a solution and the time is not so much.

**Expected behavior**
Have the possibility to migrate from AppCenter Push to  NotificationHub without have to re-install all the applications from zero but just using the in-app update provided by AppCenter.

**Comment by elamje** [comment]
I'm in a similar boat as the OP. The lack of documentation for the new MSNotificationHub is crushing the development of our application. On the android side, I am able to save a registration id to a db to target that device with a specific update, however the iOS sdk (MSNotificationHub) doesn't have the same concept. There is no clear way to get a unique device identifier that Notification Hubs can use to target the device. 

In my case, we use a backend to send out custom push notifications, so we must persist a token to the db that can be used to push to the specific device, but iOS is lacking this token, or at least any documentation on how to access this token.

**Comment by marstr** [comment]
Howdy folks, I'm truly sorry to hear that you are experiencing these frustrations with our documentation while migrating.

To keep this thread focused, I'd like to focus this discussion on the original bug mentioned by @andrekiba:

> The problem is that the NotificationHub Installation always fails in case of in-app update (OnInstallationSaveFailed).
Instead it always works re-installing the application as new (not a possible scenario at all!!).

The stack trace indicates a client-side error, which likely means that there is a bug in the underlying Android SDK.

**Comment by marstr** [comment]
I've dug into this a bit today, and here's what I've found:

It's pretty easy for me to repro getting a 400 on first launch of an Android application. This is happening because the push-service (Firebase in this case) hasn't had time to initialize and assign the device a unique identifier for cloud messaging quite yet, and is losing a race with our Installation update pipeline. However - this does almost immediately resolve itself because as soon as a device id is assigned, the `FirebaseReceiver` will update the push channel and a new Installation PUT request is sent automatically.

This is a real paper-cut, because anybody who sets an `InstallationUpdateFailureListener` will get invoked once without an indication that it is likely a temporary problem. We can start thinking about ways to fix this, at worst by deliberately losing the race. But in the meantime, I don't believe it should block you from adopting Notification Hubs.

Let me know if you are getting different details on your error. I'd definitely like to follow up and make sure there are no blockers for your adoption of Notification Hubs.

**Comment by andrekiba** [comment]
Hello and thanks in advance @marstr, after some investigations I also noticed the 400 error behaviour on first run, this triggers the OnInstallationFailed event, but then, as you said, also the OnInstallationSaved will be triggered and so I finally have the notification hub installation id that I can then persist on db.
Unfortunately this is true for new applications...or at least it seems.
In case you already have an app and you need to proceed with an in-app update to migrate from AppCenter to NotificationHub this is not true.
You have the 400 error and that’s it. OnInstallationSaved is never triggered that means no way to know the installation id. I also tried to force the saving of the installation inside the OnInstallationFailed event handler, without success.
Could you please investigate this? Because this is the blocking part.
And also, how about the expiration of the installation?

**Comment by marstr** [comment]
Interesting - that does sound like it warrants further investigation. I'll work on repro'ing this. In the meeantime @andrekiba, the exception that is being raised looks like its of type `ClientException`. That type has methods `getStatusCode` and `getResponseBody`. Mind logging those, or reading from them in a debugger and posting the error that we are responding with?

**Comment by marstr** [comment]
To your question about expiration - You are correct. If the client does not specify an expiration, the server will not apply one. The new ANH SDK acts as the client and adds one by default - but also will automatically extend that expiration date everyday when your user first launches the app.

**Comment by andrekiba** [comment]
Hello @marstr, the `ClientException` seems defined in `Xamarin.Azure.NotificationHubs.Android` correct? It is not available in Xamarin.Forms. So in which way I can retrieve what you asking for? I suppose I have to manage the hub connection inside the Xamarin.Android project and not the Forms one. Or I can inspect something from `InstallationSaveFailedEventArgs`?

About the expiration of the installation I strongly suggest to modify the behavior to avoid the expiration by default (that is exactly what the notification hub documentations says). And let the client to eventually decide to set the expiration.
This is because you have to think that some applications are meant to not be re-started every day, so the day-by-day renew is not applicable at all. Please let me know also about this topic because is another serious thing to consider.

**Comment by marstr** [comment]
Hi again @andrekiba,

You should be able to inspect the Exception that lives inside of `InstallationSaveFailedEventArgs`. Here's a screen shot of me inspecting a rejection after I deliberately add an invalid tag during initialization:

![image](https://user-images.githubusercontent.com/11203166/97764233-2d8f9580-1acb-11eb-9ebd-33d87c89c82d.png)

As you can see, `e.Exception` has a `base` of `NotificationHubException` that has properties for ResponseBody, StatusCode, etc. That's what we're looking for. Other than that, you could wrestle with your dependencies to add a using for the Android specific types and cast to the `ClientException` safely.

Regarding expiration, I hear your concern; and I agree that forcing users to launch every day would be too aggressive. While we update the expiration once a day, that expiration is pushed out by [90 days each time](https://github.com/Azure/azure-notificationhubs-android/blob/84e8d0bae65d8efd73f6c061c03d83a875ad6b05/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHubInstallationAdapter.java#L36). If that time window doesn't work for you, we did make sure that you can initialize `NotificationHub` with your own `InstallationAdapter`. So you could instantiate an adapter with a more appropriate window for your application using [this constructor](https://github.com/Azure/azure-notificationhubs-android/blob/84e8d0bae65d8efd73f6c061c03d83a875ad6b05/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHubInstallationAdapter.java#L53). Or if the `Installation` should never expire, you could easily write your own `InstallationAdapter` that doesn't add it.

**Comment by andrekiba** [comment]
OK @marstr thank you, I try to find a way to retrieve the informations you need. Sorry for the delay but unfortunately it is not easy to debug a production ready app distrubuted with AppCenter. You have someway to "simulate" the in-app update process...with the possibility to attach the debugger.

**Comment by marstr** [comment]
In the next update this information will be easier to retrieve - because it will be embedded in the Exception message already. So you're already helping make this SDK better :)

**Comment by andrekiba** [comment]
Because trying to debug a release version seems almost impossibile I did something like this:
```
static void OnInstallationSaveFailed(object sender, InstallationSaveFailedEventArgs e)
{
    AppCenterLog.Info(Constants.AmbrogioLogTag, Events.NotificationHubInstallationFailed);
    AppCenterLog.Info(Constants.AmbrogioLogTag, $@"Failed to save installation: {e.Exception.Message}");

    Ach.Instance.TrackEvent(Events.NotificationHubInstallationFailed);
    Ach.Instance.TrackError(e.Exception);

    throw e.Exception;

    try
    {
        //try to save installation again
        NotificationHub.SaveInstallation();
    }
    catch (Exception exception)
    {
        AppCenterLog.Info(Constants.AmbrogioLogTag, Events.NotificationHubInstallationRetryFailed);
        AppCenterLog.Info(Constants.AmbrogioLogTag, $@"Failed retry to save installation: {exception.Message}");
        
        Ach.Instance.TrackEvent(Events.NotificationHubInstallationRetryFailed);
        Ach.Instance.TrackError(exception);
    }
}
```
so I added the line `throw e.Exception;` and then I log the `NotificationHubException` in the Android project where I can use the type, as following
```
static void TaskSchedulerOnUnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs unobservedTaskExceptionEventArgs)
{
    foreach (var innerEx in unobservedTaskExceptionEventArgs.Exception.InnerExceptions)
    {
        if (innerEx is NotificationHubException nhe)
        {
            Ach.Instance.TrackError(nhe, new Dictionary<string, string>
            {
                {"StatusCode", nhe.StatusCode.ToString()},
                {"Responsebody", nhe.ResponseBody}
            });
        }
    }
    
    var ex = new Exception("TaskSchedulerOnUnobservedTaskException", unobservedTaskExceptionEventArgs.Exception);
    System.Diagnostics.Debug.WriteLine(ex);
    Ach.Instance.TrackError(ex);
}

static void CurrentDomainOnUnhandledException(object sender, UnhandledExceptionEventArgs unhandledExceptionEventArgs)
{
    if ((unhandledExceptionEventArgs.ExceptionObject as Exception) is NotificationHubException nhe)
    {
        Ach.Instance.TrackError(nhe, new Dictionary<string, string>
        {
            {"StatusCode", nhe.StatusCode.ToString()},
            {"Responsebody", nhe.ResponseBody}
        });
    }
    
    var ex = new Exception("CurrentDomainOnUnhandledException", unhandledExceptionEventArgs.ExceptionObject as Exception);
    System.Diagnostics.Debug.WriteLine(ex);
    Ach.Instance.TrackError(ex);
}
```
I know that re-throwing the exception I loose the stack trace but maybe we can have what we need.
Unfortunately I tested this code with 3 different installations and I'm now not able to have the InstallationFailed event anymore...I'm trying to understand what I'm doing wrong... :-(

**Comment by marstr** [comment]
> Can be something related to the fact that we have different Azure NotificationHub for dev and prod environment but they are both connected to the same FireBase app?

That shouldn't be a problem.

It may be time to transfer this over to our official Azure support channels. That way we'll be able to safely communicate your personal information and look into your troubles from our side. In the Azure portal page for your Notification Hub, you should see a button that says "New support request" that will get the process started.

**Comment by marstr** [comment]
Version 1.1.3 of the Android SDK now includes more details in the Exception message by default, and should help diagnose the problems you're running into.

**Comment by andrekiba** [comment]
Since the last time I still have the same error in case of in-app update...not if I start from a new app.

Raising the same issue on app-center side, as you suggested, they just found the time to immediately close the issue saying that you are in contact... https://github.com/microsoft/appcenter-sdk-dotnet/issues/1458
Sorry but...it is something unbelievable...the deadline for the stop of AppCenter Push is still at the end of this month...and there isn't an official package...nor an official path to follow for the migration...

In which way using your library I can put in place a custom InstallationAdapter in Xamarin Forms? ...and sincerely I still cannot understand why you want to reverse the "no-expiration" that is the default for NotificationHub based on the documentations...

**Comment by marstr** [comment]
I'm sorry, I can feel your frustration and I get it.

I'm sorry you're still encountering issues. Have you been able to retrieve the body of the 400 yet, so that I have a shot at being able to help you? In the most recent release, the server's error details should be in the exception message.

**Comment by andrekiba** [comment]
Martin I really appreciate your effort.
No, I cannot retrieve the body of the error because I can't debug the in-app update and directly use the debugger to inspect the error.
Please look at the code above where I use `TaskSchedulerOnUnobservedTaskException` and `CurrentDomainOnUnhandledException`.
Every non-catched exception goes here...I try to log useful information like the StatusCode and the ResponseBody.
But unfortunately inside AppCenter logs I do not have info about these two ...and the only idea that I have is that the exception raised by the sdk is not of type `NotificationHubException` ...otherwise I should have them.
Moreover you can see that once it fails the SDK keeps bombarding with calls that keep failing!

![image](https://user-images.githubusercontent.com/12699760/102681069-b8813800-41be-11eb-8b77-a9076488a103.png)

![image](https://user-images.githubusercontent.com/12699760/102681072-c33bcd00-41be-11eb-9f1e-7a991e1cb27f.png)


May we set up a 10 minutes call and look at this together?

**Comment by andrekiba** [comment]
I also notice that in my solution I have:
```
<PackageReference Include="Xamarin.Azure.NotificationHubs.Android" Version="1.1.1" />
<PackageReference Include="Xamarin.Android.Volley" Version="1.1.1.1" />
<PackageReference Include="Xamarin.GooglePlayServices.Base" Version="71.1610.4" />
<PackageReference Include="Xamarin.Firebase.Messaging" Version="71.1740.4" />
<PackageReference Include="Xamarin.Android.Support.v7.AppCompat" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.v4" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.Compat" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.Core.UI" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.Core.Utils" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.Fragment" Version="28.0.0.3" />
<PackageReference Include="Xamarin.Android.Support.Media.Compat" Version="28.0.0.3" />
<PackageReference Include="Xamarin.GooglePlayServices.Iid" Version="71.1600.4" />
```
but now in repo you have:
```
<PackageReference Include="Xamarin.AndroidX.AppCompat" Version="1.2.0.5" />
<PackageReference Include="Xamarin.AndroidX.Migration" Version="1.0.8" />
<PackageReference Include="Xamarin.Google.Android.DataTransport.TransportBackendCct" Version="2.3.0" />
<PackageReference Include="Xamarin.Google.Android.DataTransport.TransportRuntime" Version="2.2.3" />
<PackageReference Include="Xamarin.Firebase.Messaging" Version="120.1.7" />
<PackageReference Include="Xamarin.GooglePlayServices.Iid" Version="117.0.0" />
<PackageReference Include="Xamarin.GooglePlayServices.Base" Version="117.2.1" />
<PackageReference Include="Xamarin.Android.Volley" Version="1.1.1.1" />
<PackageReference Include="Xamarin.Azure.NotificationHubs.Android" Version="1.1.3" />
```

You told me that `Xamarin.Azure.NotificationHubs.Android` was updated...and ok...the others are only dependencies??

**Comment by andrekiba** [comment]
Finally I moved everything directly in the Xamarin.Android project (not in the Forms one as before) and here I can see the error that I have only in case off update of an existing app (if I start from a fresh installation it works great)

![Schermata 2020-12-20 alle 08 49 32](https://user-images.githubusercontent.com/12699760/102708120-d6bf6480-42a0-11eb-933f-76320a3473ca.png)

I also want to say that in Xamarin.Android the bombarding of the the same error message is not present (so there is an error somewhere in you Forms binding library??).
And the constructor of `InstallationAdapter` that you mentioned is private...

**Comment by MyotisX** [comment]
> **Describe the bug**
> Seems impossibile to migrate devices from AppCenter Push to NotificationHub without re-installing the application.

> **Expected behavior**
> Have the possibility to migrate from AppCenter Push to NotificationHub without have to re-install all the applications from zero but just using the in-app update provided by AppCenter.

@andrekiba @marstr 
I'm experiencing the same issue. Has anyone made progress on this?

**Comment by andrekiba** [comment]
unfortunately no :-( ...

**Comment by marstr** [comment]
Ah, okay. Now we're getting somewhere. This error is what will be presented if Firebase fails to initialize. I've heard some other reports recently about the Release profile not working right, and it coming down to a missing hint to a package minimizer that prevents it from stripping out pieces we need. I'll work on a work around for that, and publish it here to see if it's the same issue.

**Comment by marstr** [comment]
I just mentioned in another thread - but one explanation for why this error is occurring is that the release profile is overly aggressive, and removes class definitions you actually need.

To see if this is your problem, you can edit your Release profile to not do any shrinking. If everything "just works" then we'll be able to confirm that this is the issue. To fix it, we'll need to telegraph to the shrinker that it shouldn't remove those class definitions by ensuring that the proguard configuration is setup to keep everything in the `com.microsoft.windowsazure.messaging.notificationhubs` namespace. There's some documentation [here](https://docs.microsoft.com/en-us/xamarin/android/deploy-test/release-prep/proguard?tabs=windows) that I've found helpful.

**Comment by andrekiba** [comment]
@marstr seems this is not the problem...
The "Release" profile is the same for both the cases (starting from a fresh installation or starting from an old version of the application with an in-app update), the builded apk is exactly the same apk.
Another time...this is a big problem...do you have another idea?

![image](https://user-images.githubusercontent.com/12699760/103271947-e08d4a00-49bb-11eb-93d8-cc9d5e2bdaec.png)

**Comment by andrekiba** [comment]
Could the problem be related to the fact that in case of update of an already existing app you already have (I suppose) a Firebase registration to uniquely identify the device?
The registration was made by the AppCenter Push sdk [https://docs.microsoft.com/en-us/appcenter/sdk/push/xamarin-android](url).

**Comment by andrekiba** [comment]
Hello @marstr do we have any news on this?
We are actually without the service, the only way seems to re-install the app from the scratch.

**Comment by tharindup** [comment]
@andrekiba after the update once you opened the app for the first time, did you try closing the app and open again?

We are following a similar pattern like you and got a similar error. When we update the old app (which was using **AppCenter**) to the **Notification Hub** version and sent notifications using `$InstallationId:{installationId}` tag format, the app did not receive any notification. When we checked the Hub's registrations the app hasn't registered in the hub as well. That means even when we call the NotificationHub.Start method at the app startup the app did not register with the Notification Hub. 
This happens only when we open the app just after the update. 

Opening the app for the first time after the update, and close the app from the stack and reopen again registered the app in the NotificationHub, and started receiving notifications as expected.

**Comment by andrekiba** [comment]
In our case no, reopening the app doesn't solve...

**Comment by marstr** [comment]
Howdy folks, I just wanted to update thread. I'm working on reproducing this locally. So far, native Java apps don't exhibit this problem. I'll report back with Xamarin/Java bindings soon.

**Comment by marstr** [comment]
Alright - I was able to repro the problem. I'll keep digging and hopefully be able to fix this.

**Comment by andrekiba** [comment]
Thank you @marstr I hope we can have a solution soon.

**Comment by marstr** [comment]
Alright! I have the problem figured out. Basically, in our situation an in-app update doesn't create a new Firebase Instance ID. This means that our FirebaseReceiver is never woken up, which means it never seeks the device identifier, and never hands it to the NotificationHub class or our servers.

I was able to successfully wake up the receiver by simply sending a broadcast Notification from App Center after I had installed the new version. After that, I can receiver notifications from Azure Notification Hubs.

Obviously, that's not a perfect work around. I'm investigating for more programatic solutions now.

**Comment by marstr** [comment]
Alright @andrekiba & @MyotisX, 

Please try adding the following line, immediately following your call to `NotificationHub.Start`:

_C#, Xamarin_
``` C#
this.StartService(new Intent(this, typeof(FirebaseReceiver)));
```
_Java_
``` Java
// import com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver;
this.startService(new Intent(this, FirebaseReceiver.class));
```

This should act as a work around, and I'll look into ways to avoiding the lazy-start that Android puts on this service, so that other customers aren't impacted in the future.

**Comment by marstr** [comment]
Alright, there's a PR that also addresses the lazy-start now: https://github.com/Azure/azure-notificationhubs-android/pull/175

We'll make sure that the next version of the SDK doesn't have this problem, and that it is released ASAP.

**Comment by andrekiba** [comment]
ok @marstr thank you, for now I try to use the StartService as a work around and see if it works.
Then, when the https://github.com/Azure/azure-notificationhubs-android/pull/175 will be solved,
will it also be integrated in this Xamarin Forms project?
When we will have an official nuget package?

**Comment by andrekiba** [comment]
do we still need these lines in the manifest?
```
<receiver android:name="com.google.firebase.iid.FirebaseInstanceIdInternalReceiver" android:exported="false" />
<receiver android:name="com.google.firebase.iid.FirebaseInstanceIdReceiver" android:exported="true" android:permission="com.google.android.c2dm.permission.SEND">
    <intent-filter>
        <action android:name="com.google.android.c2dm.intent.RECEIVE" />
        <action android:name="com.google.android.c2dm.intent.REGISTRATION" />
        <category android:name="${applicationId}" />
    </intent-filter>
</receiver>
```
I try to test your work around better but with a first try seems that the app crashes exactly when it tries to start the FirebaseReciever...

**Comment by marstr** [comment]
@andrekiba 
> do we still need these lines in the manifest?

No - The SDK itself declares everything it needs in a manifest that it provides, and will automatically get merged with yours at compile time.

>Then, when the Azure/azure-notificationhubs-android#175 will be solved, will it also be integrated in this Xamarin Forms project? 

Once we publish a new copy (wheels are turning there), new Xamarin bindings will be generated and published. I'll make sure that the Xamarin Form project adopts it.

> When we will have an official nuget package?

A totally understandable question. It is on our road-map, but I can't commit to a time when it will be delivered yet.

**Comment by andrekiba** [comment]
@marstr I can confirm that your trick crashes the app...
```
NotificationHub.SetListener(new NotificationListener());
NotificationHub.SetInstallationSavedListener(new InstallationSavedListener());
NotificationHub.SetInstallationSaveFailureListener(new InstallationSaveFailedListener());
NotificationHub.Start((Android.App.Application)Android.App.Application.Context, Constants.NotificationHubName, Constants.NotificationHubConnectionString);
--> StartService(new Intent(this, typeof(FirebaseReceiver)));
```

Adding the line to try to start the FirebaseReceiver leads to this exception.

```
Java.Lang.RuntimeException: Unable to create service com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver: java.lang.IllegalStateException: Default FirebaseApp is not initialized in this process net.elfo.ambrogio. Make sure to call FirebaseApp.initializeApp(Context) first. ---> Java.Lang.IllegalStateException: Default FirebaseApp is not initialized in this process net.elfo.ambrogio. Make sure to call FirebaseApp.initializeApp(Context) first.
   --- End of inner exception stack trace ---
  --- End of managed Java.Lang.RuntimeException stack trace ---
java.lang.RuntimeException: Unable to create service com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver: java.lang.IllegalStateException: Default FirebaseApp is not initialized in this process net.elfo.ambrogio. Make sure to call FirebaseApp.initializeApp(Context) first.
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3352)
	at android.app.ActivityThread.-wrap4(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1680)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6518)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
Caused by: java.lang.IllegalStateException: Default FirebaseApp is not initialized in this process net.elfo.ambrogio. Make sure to call FirebaseApp.initializeApp(Context) first.
	at com.google.firebase.FirebaseApp.getInstance(com.google.firebase:firebase-common@@16.1.0:245)
	at com.google.firebase.iid.FirebaseInstanceId.getInstance(Unknown Source:1)
	at com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver.onCreate(FirebaseReceiver.java:52)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3342)
	... 8 more
```

**Comment by marstr** [comment]
This is likely an independent issue that you're running into. This can happen when Google Service aren't being applied to your application, or if your application ID doesn't appear in your `google-services.json`.

Side note, NotificationHub.Start should be called with an Application, not an Application Context.

I am sorry to hear you're still having trouble. I do think it's important to note for future people reading the thread, that we've confirmed that the workaround has worked for other customers and that they are receiving notifications as expected now. Also, we've published the Azure Notification Hubs Android SDK version 1.1.4, and the Xamarin binding is on the way. Once you adopt that version you won't need the workaround anymore.

**Comment by andrekiba** [comment]
Hello @marstr,
yes I know about the Application, the line

`NotificationHub.Start((Android.App.Application)Android.App.Application.Context, Constants.NotificationHubName, Constants.NotificationHubConnectionString);`

was just a test, if I use the

`NotificationHub.Start(this.Application Constants.NotificationHubName, Constants.NotificationHubConnectionString);`

I have the same error calling the `StartService(new Intent(this, typeof(FirebaseReceiver)));` just after.

The `google-services.json` inside the Android project is exactly the same since more or less two years ago when AppCenter Push were introduced...
I do not want to post it here, I send you a message.

**Comment by marstr** [comment]
Version 1.1.4.1 of the bindings for the Android SDK is now available on NuGet and should allow you to try a solution which doesn't involve the work around. However, if Firebase isn't initializing, you'll continue to see the same error. The stack trace will just be in a slightly different location. I can't get into details of other customers, but I can confirm that I've ensured that the issue with App Center Distribute in-app updates is now fixed, and others have verified the solution.

For logistical reason, I need to limit this scope of this thread. I've been using it to track the progress on ensuring in-app updates to ANH from former App Center Push applications worked. If 1.1.4.1 throws the `llegalStateException: Default FirebaseApp is not initialized in this process.` exception again, let's open a new thread to track that investigation.
